opensearch-project/OpenSearch
# [Tiered Caching] Enabling serialization for IndicesRequestCache key o…

### Description
As part of enabling tiered caching for RequestCache [here](https://github.com/opensearch-project/OpenSearch/issues/10024), we need the ability to serialize its key/value to be able to offload to disk.

RequestCache key is a composite one comprising query object, shardEntity and IndexReader.CacheKey(used to invalidate when refresh/merge happens). IndexReader.CacheKey isn't serializable as it is defined under lucene library, is final (can't be extended) and doesn't contain any fields etc.

This change wraps CacheHelper and CacheKey in **OpenSearchDirectoryReader** so that we can attach a unique string id to CacheKey instead of using object itself which isn't serializable.

Similarly for shardEntity, this change adds a way to make it writeable. We are using OpenSearch existing streamInput/output mechanism for serialization.



### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/10308


### Alternate approach
Change internal lucene IndexReader.CacheKey itself to attach a uniqueID string which instead can be used.

### Testing

- Manual testing
Ran local opensearch server.
1. Create shard and Indexed documents. Refresh.
2. Fire a query 1 and verify the results. Request cache (miss count: 1, hit count: 0)
3. Fire query 1 above again. Request cache (miss count: 1, hit count: 1)
4. Update documents.
5. Fire query 1 again. Request cache (miss count: 1, hit count: 2) (As data isn't refreshed yet)
6.  Refresh
7.  Fire query 1 again. Request cache (miss count: 2, hit count: 2) (due to refresh cacheKey changes, making it a miss)

- Integ test/Unit test
Added new integ/UT. Verified existing ones.

### Check List

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
